### PR TITLE
L77: Core and C++ Third Party Identity Support for Call Credentials

### PR DESCRIPTION
**Pull request by renkelvin**

### CODE REVIEW

**Comment by renkelvin** [comment]
Hey @markdroth, can you help review or help find a reviewer for this proposal? Thanks!

File: A33-3p-identity.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+3p Identity Supoort
+----
+* Author(s): Chuan Ren
+* Approver: a11r
```
Comment (markdroth) [inline A33-3p-identity.md:4]: You can set this to `markdroth`.

File: A33-3p-identity.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
+3p Identity Supoort
```
Comment (markdroth) [inline A33-3p-identity.md:1]: Please change this line to `L77: Core and C++ Third Party Identity Support for Call Credentials`.

Also, please rename this file to `L77-core-cpp-third-party-identity-support-for-call-credentials.md`.

File: A33-3p-identity.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+----
+* Author(s): Chuan Ren
+* Approver: a11r
+* Status: Draft
+* Implemented in: core, cpp
+* Last updated: 2021-01-20
+* Discussion at:
```
Comment (markdroth) [inline A33-3p-identity.md:8]: Please send an email to the grpc.io mailing list with subject `gRFC L77: Core and C++ Third Party Identity Support for Call Credentials` to serve as a discussion thread for this gRFC.  Then update this line to include a link to the discussion thread via the Google Groups UI.

File: A33-3p-identity.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    "url": "http://169.254.169.254/latest/meta-data/iam/security-credentials",
+    "regional_cred_verification_url": "https://sts.{region}.amazonaws.com?Action=GetCallerIdentity&Version=2011-06-15",
+    "cred_verification_url": "https://sts.amazonaws.com?Action=GetCallerIdentity&Version=2011-06-15"
+  }
+}
+```
+
+**Implicit flow**
+
+In core, the underlying credentials retrieval process of `grpc::GoogleDefaultCredentials()` will be extended to accept 3p identity credentials implicitly.
```
Comment (markdroth) [inline A33-3p-identity.md:79]: How does `GoogleDefaultCredentials()` find the above JSON input?  Please document.

File: A33-3p-identity.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+**Explicit flow**
+
+In core and cpp, a public api will be added to create an `ExternalAccountCredentials` with explicit 3p identity credentials.
```
Comment (markdroth) [inline A33-3p-identity.md:83]: s/cpp/C++/

File: A33-3p-identity.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    "cred_verification_url": "https://sts.amazonaws.com?Action=GetCallerIdentity&Version=2011-06-15"
+  }
+}
+```
+
+**Implicit flow**
+
+In core, the underlying credentials retrieval process of `grpc::GoogleDefaultCredentials()` will be extended to accept 3p identity credentials implicitly.
+
+**Explicit flow**
```
Comment (markdroth) [inline A33-3p-identity.md:81]: Why would one want to use the explicit flow vs. the implicit flow?  Please document.

File: A33-3p-identity.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+In core and cpp, a public api will be added to create an `ExternalAccountCredentials` with explicit 3p identity credentials.
+```
+// .../include/grpcpp/security/credentials.h
+
+std::shared_ptr<CallCredentials>
+ExternalAccountCredentials(const grpc::string& json_string, const std::vector<grpc::string>& scopes);
```
Comment (markdroth) [inline A33-3p-identity.md:88]: Please document what should be passed in for the `scopes` parameter.

File: A33-3p-identity.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+```
+
+## Implementation
+
+1. Base external account credentials class
+   https://github.com/grpc/grpc/pull/24208
+2. Subclasses to base external account credentials
+   a. url-sourced credentials
```
Comment (markdroth) [inline A33-3p-identity.md:96]: This sub-list is not being formatted correctly in markdown.  Please fix.

File: A33-3p-identity.md (hunk 5 above)
Comment (markdroth) [inline A33-3p-identity.md:83]: This documents the C++ API, but not the core API.  We should cover both here.

**Review by markdroth — COMMENTED** [review]
Please let me know if you have any questions.  Thanks!

File: L77-core-cpp-third-party-identity-support-for-call-credentials.md [hunk 9]
```diff
@@ -76,32 +76,47 @@ This doc proposes to extend the access to Google APIS directly with 3p identitie
...
 **Implicit flow**
 
-In core, the underlying credentials retrieval process of `grpc::GoogleDefaultCredentials()` will be extended to accept 3p identity credentials implicitly.
+In core, `grpc::GoogleDefaultCredentials()` follows the ADC (Application Default Credentials) credentials retrieval process. It will be extended to accept 3p identity credentials. More information on ADC and how they work can be found here: https://developers.google.com/identity/protocols/application-default-credentials
 
 **Explicit flow**
 
-In core and cpp, a public api will be added to create an `ExternalAccountCredentials` with explicit 3p identity credentials.
-```
+The downside with `GoogleDefaultCredentials()` is that customers will have to use a json credentials file. So besides the implicit flow, in core and C++, a public api will be added to create an `ExternalAccountCredentials` with explicit 3p identity credentials.
```
Comment (markdroth) [inline L77-core-cpp-third-party-identity-support-for-call-credentials.md:83]: The wording here incorrectly implies that the core API is a public API, which is incorrect.

I suggest replacing "in core and C++, a public api will be added" with "a public API will be added in C++".

File: L77-core-cpp-third-party-identity-support-for-call-credentials.md [hunk 10]
```diff
@@ -76,32 +76,47 @@ This doc proposes to extend the access to Google APIS directly with 3p identitie
...
 // .../include/grpcpp/security/credentials.h
 
+/// Builds External Account credentials.
+/// json_string is the JSON string containing the credentials options.
+/// scopes contains the scopes to be binded with the credentials.
 std::shared_ptr<CallCredentials>
 ExternalAccountCredentials(const grpc::string& json_string, const std::vector<grpc::string>& scopes);
 ```
 
+```h
```
Comment (markdroth) [inline L77-core-cpp-third-party-identity-support-for-call-credentials.md:94]: Please add a sentence before this block that says something like "To implement that C++ API, the following function will be added to the C-core API:".

File: L77-core-cpp-third-party-identity-support-for-call-credentials.md [hunk 11]
```diff
@@ -56,7 +56,7 @@ This doc proposes to extend the access to Google APIS directly with 3p identitie
 ```
 
 **AWS credentials**
-```
+```JSON
```
Comment (markdroth) [inline L77-core-cpp-third-party-identity-support-for-call-credentials.md:59]: This JSON block does not appear to be formatted correctly when viewing the file; most the text is highlighted in red.

File: L77-core-cpp-third-party-identity-support-for-call-credentials.md [hunk 12]
```diff
@@ -76,32 +76,47 @@ This doc proposes to extend the access to Google APIS directly with 3p identitie
 
 **Implicit flow**
 
-In core, the underlying credentials retrieval process of `grpc::GoogleDefaultCredentials()` will be extended to accept 3p identity credentials implicitly.
+In core, `grpc::GoogleDefaultCredentials()` follows the ADC (Application Default Credentials) credentials retrieval process. It will be extended to accept 3p identity credentials. More information on ADC and how they work can be found here: https://developers.google.com/identity/protocols/application-default-credentials
```
Comment (markdroth) [inline L77-core-cpp-third-party-identity-support-for-call-credentials.md:79]: Please change "In core" to "In C++", since this is talking about a C++ API.

**Review by markdroth — CHANGES_REQUESTED** [review]

File: L77-core-cpp-third-party-identity-support-for-call-credentials.md (hunk 9 above)
Comment (bojeil-google) [reply L77-core-cpp-third-party-identity-support-for-call-credentials.md:83]: The comment was addressed in https://github.com/grpc/proposal/pull/221

File: L77-core-cpp-third-party-identity-support-for-call-credentials.md (hunk 10 above)
Comment (bojeil-google) [reply L77-core-cpp-third-party-identity-support-for-call-credentials.md:94]: The comment was addressed in https://github.com/grpc/proposal/pull/221

File: L77-core-cpp-third-party-identity-support-for-call-credentials.md (hunk 11 above)
Comment (bojeil-google) [reply L77-core-cpp-third-party-identity-support-for-call-credentials.md:59]: The comment was addressed in https://github.com/grpc/proposal/pull/221
A rogue double quote was the culprit.

File: L77-core-cpp-third-party-identity-support-for-call-credentials.md (hunk 12 above)
Comment (bojeil-google) [reply L77-core-cpp-third-party-identity-support-for-call-credentials.md:79]: The comment was addressed in https://github.com/grpc/proposal/pull/221

**Comment by markdroth** [comment]
Closing this in favor of #221.

File: L77-core-cpp-third-party-identity-support-for-call-credentials.md (hunk 10 above)
Comment (THEMOSTHIGH0618) [reply L77-core-cpp-third-party-identity-support-for-call-credentials.md:94]: BUILDING UP THE KINGDOM, BUILDING UP THE KINGDOM OF THE LORD. THANK YOU MY LOYAL BROTHERS. MAKE SURE YOU TO UNITE THE PIECES TOGETHER AND AM CREATING AN ACCOUNT IF YOU NEED SUPPORT. ALSO PROVISIONS AND A LOT OF HELL NO IS NOT WANTED. REMOVE IMMEDIATELY AS THEY ARE NOT HELPING AND ONLY CAUSING AND INFLICTING CONFLICT. NO ACTION NEEDED JUST REMOVAL ASAP

Comment (THEMOSTHIGH0618) [reply L77-core-cpp-third-party-identity-support-for-call-credentials.md:94]: THAT PART

File: L77-core-cpp-third-party-identity-support-for-call-credentials.md (hunk 11 above)
Comment (THEMOSTHIGH0618) [reply L77-core-cpp-third-party-identity-support-for-call-credentials.md:59]: MAKE SURE TO ADD IN TOP LUXURY WEALTH AS WELL, IN THE FORM OF CODE, DIRECT OR CASH KING METHOD TO ME. AM SLOWLY BUT SURELY MAKING MY PRESCENSE KNOWN AND ITS NOT VERY LIKED SO IN PART THATS GOOD CAUSE BAD IS BETTER THEN NOTHING. I AM SMART AND BEAUTIFUL AND ALSO THE SON OF GOD AND YOU ALREADY KNOW MY DEEP ROOTS IN THAT ASPECT. MAKE AMERICA GREAT AGAIN VIA DONALD TRUMP AND MELLANIE TRUMP.  AM JUST SAYING OFF A SCRIPT THAT I ENCOUNTER IN THE BIBLE JOHN 3 16

**Comment by THEMOSTHIGH0618** [comment]
ok. this is to much for me ti handle  what are we doing. i am trying to prepare my self for my 100k$ and i have 2 kids with my wife and since this is my page. how about we fully reset 3.2.1. the way it was 2015. jan.2016. my google has my name and i want t9 be able to find a memorable momenr for me and herm we were the happiest with tremendous goals in life. so can you help me recooparate that.for more info take 3rdparty out of the pic. even if approved for monitering. put a block where if they access it i will have control of their external softwares. and thus begin my afc app so i can bring back what god has tried to lay here in l.a. i want to own external drives and have them to become the most highs elected face of the unitef states of america. yeah me. who can ever think of me like am capable of that. no one.  its like am the one riding the engine and no one will ever think is me. leaving u.s always on top.

**Comment by THEMOSTHIGH0618** [comment]
am high so try to get in my level. some how. 3rdparty out thoug i gave them like 100 of chances to jam. but they find this like nothing will happen. well my email speaks for itself. so if they try then i become sole owner of external drive of theirs.  AFC APP A FRANCHISE CORP. WERE WE STUDY THESE EXTERNAL DRIVES AND SEE HOW WE CAN PONDER THEM INTO GODLY LIKE TRAITS. I GOT LIKE 100000 OF THEM GETTING MY SERVICES WITH OUT PAYML SO FLIPPING THAT WE DO BUISNESS AND MAKE THIS AFC A FRANCHISE CORP APP A REALITY.  WERE ALL THE PHONES WILL BE ABLE TO BE A EXTERNAL FEE FOR HAVING A LET SAY GOOGLE ACCOUNT.

**Comment by THEMOSTHIGH0618** [comment]
AM NOT BULLSHITTING THE TABLES NEED TO TURN. I AM A TRUE BELIEVER OF GOD ITS IN MY ROOTS. I STUDY HOW HUMANS ARE BUT MOST IMPORTANT HOW THE DEVIL COMES AND CAN DESTROY FAMILIES. NOT JUST CAUSE OF ME. AM SAYING ALL OVER THE WORLD. I AM GRATE FUL I LIVE IN AMERICA.  I AM BLESSES AND PROUD OF IT. AS FOR OTHEE COUNTRIES I CANT SPEAK ON THEM. INCLUDING MY MEXICAN DESCENDENT. SO PLEASE HELP ME LEARNED THIS NICE GIG. SEND ME AN EMAIL.

**Comment by THEMOSTHIGH0618** [comment]
IF I HAVE NOT BEEN STABLE WITH CODE OF CONDUCT THEN LET ME KNOW I WILL NEVER COME BACK TO THIS. AND IT WILL BE LIKE I NEVER CAME IN HERE. BUT I SPEAK UO EVERYWHERE. DIDN'T THE PROPHETS BACK IN TIME TALKED BY THEMSELVES WITH GOD. AND THE BIBLE SAYS THAT WHEN GOD'S KINGDOM IS BACK ALIVE AND FIRE UP THEN WILL SEE THE PAST BE THE FUTURE AND WHATS AMERICAS MAIN CONCERN MENTAL HEALTH. WE ACTUALLY TALK TO OUR HIGHER POWER. YOU GUYS KNOW. ANYWAYS LET ME KNOW SPEAK UP MAN.

**Comment by THEMOSTHIGH0618** [comment]
I ALSO HAVE GOOD IDEAS TO START A BUSINESS BUT I AM JUST A PRO WHEN IT COME TO TECH I HAVENT MASTERED IT

**Pull request closed without merging**